Use CGStyle-based filters for saturate() and hue-rotate() - dropped frames on fakecake.org

smfr opened the pull request:
#### e664737f1d2f8c6567134999415c930933336cd1
<pre>
Use CGStyle-based filters for saturate() and hue-rotate() - dropped frames on fakecake.org
<a href="https://bugs.webkit.org/show_bug.cgi?id=293212">https://bugs.webkit.org/show_bug.cgi?id=293212</a>
<a href="https://rdar.apple.com/148338705">rdar://148338705</a>

Reviewed by Sam Weinig.

The code in `FEColorMatrix::supportedFilterRenderingModes()` caused fallback to `FilterRenderingMode::Software`
for `hue-rotate()` and `saturate()` filters, because we didn&apos;t have code that converted these to a color matrix.
However, that code exists in ColorMatrix, so we can use it.

I attempted to have GraphicsColorMatrix own a ColorMatrix&lt;5, 4&gt;, but that isn&apos;t easily IPC-encodable. Instead,
add a `ColorMatrix` constructors that takes a span, and make the variadic constructor limited to numeric types.

* LayoutTests/css3/filters/effect-blur-hw.html:
* LayoutTests/css3/filters/effect-brightness-hw.html:
* LayoutTests/css3/filters/effect-combined.html:
* LayoutTests/css3/filters/effect-grayscale-hw.html:
* LayoutTests/css3/filters/effect-hue-rotate-hw.html:
* LayoutTests/css3/filters/effect-hue-rotate.html:
* LayoutTests/css3/filters/effect-invert-hw.html:
* LayoutTests/css3/filters/effect-opacity-hw.html:
* LayoutTests/css3/filters/effect-saturate.html:
* LayoutTests/css3/filters/effect-sepia-hw.html:
* LayoutTests/css3/filters/invalid-reference-filter-in-chain.html:
* LayoutTests/css3/filters/resources/filter-helpers.css: Use image-rendering: crisp-edges to reduce noise on iOS where images get scaled.
* LayoutTests/fast/filter-image/filter-image.html:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ColorMatrix.h:
(WebCore::ColorMatrix::ColorMatrix):
(WebCore::ColorMatrix::data const):
(WebCore::to5x4Matrix):
* Source/WebCore/platform/graphics/GraphicsStyle.h:
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterRenderingModes const):
(WebCore::FEColorMatrix::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FilterRenderingMode.cpp: Copied from Source/WebCore/platform/graphics/filters/FilterRenderingMode.h.
(WebCore::operator&lt;&lt;): Make FilterRenderingMode streamable.
* Source/WebCore/platform/graphics/filters/FilterRenderingMode.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create): Log the rendering mode.

Canonical link: <a href="https://commits.webkit.org/295269@main">https://commits.webkit.org/295269@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/306eabae697c6133b3321f2dbc7a154496936f5f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109794 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55254 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106623 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32838 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79416 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107589 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19191 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94388 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59728 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18971 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12463 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54627 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88670 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12515 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112184 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31744 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23367 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88495 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32108 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90622 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88114 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22448 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33017 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10786 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31671 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37012 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31463 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34802 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33023 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/ColorMatrix.h:
```diff
@@ -35,7 +35,12 @@ template<typename, size_t> struct ColorComponents;
-    template<typename ...Ts>
+    explicit constexpr ColorMatrix(std::span<const float, RowCount * ColumnCount> s)
+    {
+        std::copy(s.begin(), s.end(), m_matrix.begin());
```
```suggestion
        std::ranges::copy(s, m_matrix.begin());
```

weinig commented on Source/WebCore/platform/graphics/ColorMatrix.h:
```diff
@@ -50,12 +55,24 @@ class ColorMatrix {
         return m_matrix[(row * ColumnCount) + column];
     }
 
+    const std::array<float, 20>& data() const { return m_matrix; }
```
```suggestion
    std::span<const float, RowCount * ColumnCount> data() const { return m_matrix; }
```

I am not entirely sure if the conversion to span here works without and fiddling. But it would be good to avoid leaking the std::array detail and using `RowCount * ColumnCount` instead of 20.

weinig commented on Source/WebCore/platform/graphics/ColorMatrix.h:
```diff
@@ -50,12 +55,24 @@ class ColorMatrix {
     std::array<float, RowCount * ColumnCount> m_matrix;
 };
 
+constexpr ColorMatrix<5, 4> to5x4Matrix(const ColorMatrix<3, 3>& matrix)
```
This is fine, but I do think this could be made generic and into a constructor, maybe something like:

```c++
template<size_t ColumnCount, size_t RowCount>
template<size_t FromColumnCount, size_t FromRowCount>
constexpr ColorMatrix<ColumnCount, RowCount>::ColorMatrix(const ColorMatrix<FromColumnCount, FromRowCount>& from)
{
    static_assert(ColumnCount >= FromColumnCount);
    static_assert(RowCount >= FromRowCount);

    for (size_t row = 0; row < RowCount; ++row) {
        for (size_t column = 0; column < ColumnCount; ++column) {
            if (row < FromRowCount && column < FromColumnCount)
                m_matrix[(row * ColumnCount) + column] = from.at(row, column);
            else if (row == column)
                m_matrix[(row * ColumnCount) + column] = 1.0f;
            else
                m_matrix[(row * ColumnCount) + column] = 0.0f;
        }
    }
}
```

shallawa replied:
This generic solution will be used for only one case. The ColorMatrix is either 3 x 3 or 5 x 4 and the conversion has to happen from 3 x3 to 5 x 4. So I think it is okay to simplify and specialize the solution here. I would suggest template specialized casting operator instead:

```
template<size_t ColumnCount, size_t RowCount>
class ColorMatrix {
public:
    ...
    template<size_t ToColumnCount, size_t ToRowCount>
    constexpr operator ColorMatrix<ToColumnCount, ToRowCount>() const;
    ...
};

template <> template <> constexpr ColorMatrix<3, 3>::operator ColorMatrix<5, 4>() const
{
    return ColorMatrix<5, 4> {
        at(0, 0), at(0, 1), at(0, 2), 0, 0,
        at(1, 0), at(1, 1), at(1, 2), 0, 0,
        at(2, 0), at(2, 1), at(2, 2), 0, 0,
        0, 0, 0, 1, 0
    };
}
```

weinig commented on Source/WebCore/platform/graphics/GraphicsStyle.h:
```diff
@@ -66,7 +66,7 @@ struct GraphicsGaussianBlur {
 
 struct GraphicsColorMatrix {
-    std::array<float, 20> values;
+    std::array<float, 20> values; // Really a ColorMatrix<5, 4> but that's not encodable.
```
Is there anything stopping ColorMatrix from being encodable?

smfr committed:
Use CGStyle-based filters for saturate() and hue-rotate() - dropped frames on fakecake.org
https://bugs.webkit.org/show_bug.cgi?id=293212
rdar://148338705

Reviewed by Sam Weinig.

The code in `FEColorMatrix::supportedFilterRenderingModes()` caused fallback to `FilterRenderingMode::Software`
for `hue-rotate()` and `saturate()` filters, because we didn't have code that converted these to a color matrix.
However, that code exists in ColorMatrix, so we can use it.

I attempted to have GraphicsColorMatrix own a ColorMatrix<5, 4>, but that isn't easily IPC-encodable. Instead,
add a `ColorMatrix` constructors that takes a span, and make the variadic constructor limited to numeric types.

* LayoutTests/css3/filters/effect-blur-hw.html:
* LayoutTests/css3/filters/effect-brightness-hw.html:
* LayoutTests/css3/filters/effect-combined.html:
* LayoutTests/css3/filters/effect-grayscale-hw.html:
* LayoutTests/css3/filters/effect-hue-rotate-hw.html:
* LayoutTests/css3/filters/effect-hue-rotate.html:
* LayoutTests/css3/filters/effect-invert-hw.html:
* LayoutTests/css3/filters/effect-opacity-hw.html:
* LayoutTests/css3/filters/effect-saturate.html:
* LayoutTests/css3/filters/effect-sepia-hw.html:
* LayoutTests/css3/filters/invalid-reference-filter-in-chain.html:
* LayoutTests/css3/filters/resources/filter-helpers.css: Use image-rendering: crisp-edges to reduce noise on iOS where images get scaled.
* LayoutTests/fast/filter-image/filter-image.html:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ColorMatrix.h:
(WebCore::ColorMatrix::ColorMatrix):
(WebCore::ColorMatrix::data const):
(WebCore::to5x4Matrix):
* Source/WebCore/platform/graphics/GraphicsStyle.h:
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterRenderingModes const):
(WebCore::FEColorMatrix::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FilterRenderingMode.cpp: Copied from Source/WebCore/platform/graphics/filters/FilterRenderingMode.h.
(WebCore::operator<<): Make FilterRenderingMode streamable.
* Source/WebCore/platform/graphics/filters/FilterRenderingMode.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create): Log the rendering mode.

Canonical link: https://commits.webkit.org/295269@main

aproskuryakov commented on LayoutTests/css3/filters/effect-blur-hw.html:
```diff
@@ -1,7 +1,7 @@
 <html>
 <head>
-    <meta name="fuzzy" content="maxDifference=0-54; totalPixels=0-61800">
+    <meta name="fuzzy" content="maxDifference=0-99; totalPixels=0-61800">
```
Not new in this PR, but do these tests actually verify anything for real? With so much fuzziness, it seems like the actual results could bear little semblance to the expectation.
